test/shared/src/main/scala/zio/test/BoolAlgebra.scala

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+/**
+ * A `BoolAlgebra[A]` is a description of logical operations on values of type
+ * `A`.
+ */
+sealed trait BoolAlgebra[+A] extends Product with Serializable { self =>
+  import BoolAlgebra._
+
+  /**
+   * Returns a new result that is the logical conjunction of this result and
+   * the specified result.
+   */
+  final def &&[A1 >: A](that: BoolAlgebra[A1]): BoolAlgebra[A1] =
+    both(that)
+
+  /**
+   * Returns a new result that is the logical disjunction of this result and
+   * the specified result.
+   */
+  final def ||[A1 >: A](that: BoolAlgebra[A1]): BoolAlgebra[A1] =
+    either(that)
+
+  /**
+   * Returns a new result that is the logical negation of this result.
+   */
+  final def unary_! : BoolAlgebra[A] =
+    negate
+
+  /**
+   * Returns a new result, with all values mapped to the specified constant.
+   */
+  final def as[B](b: B): BoolAlgebra[B] =
+    map(_ => b)
+
+  /**
+   * A named alias for `&&`.
+   */
+  final def both[A1 >: A](that: BoolAlgebra[A1]): BoolAlgebra[A1] =
+    and(self, that)
+
+  /**
+   * A named alias for `||`.
+   */
+  final def either[A1 >: A](that: BoolAlgebra[A1]): BoolAlgebra[A1] =
+    or(self, that)
+
+  /**
+   * If this result is a success returns `None`. If it is a failure returns a
+   * new result containing all failures that are relevant to this result being
+   * a failure.
+   */
+  final def failures: Option[BoolAlgebra[A]] =
+    fold[Either[BoolAlgebra[A], BoolAlgebra[A]]](a => Right(success(a)))(
```

Comment:
Do you need to go through `Either` here?

Reply:
I think we do to deal with the `not` case. If we don't use `Either` then `notCase` turns `Some` into `None`, but what does it do with `None`? We've lost information by only using the option so for example if we had `not(not(a))` we could't recover the original value, whereas `a.swap.swap == a`. Open to more elegant ways for achieving that but I think we definitely need some intermediate structure and `Either` seems to work relatively well.

Reply:
Of course 👍